Add granular settings defaults (`defaults.settings.{local, defined, all}`)

shivaraj-bh opened the pull request:
resolves #261 

`defaults.settings.default-{current, defined, all}` replaces `defaults.settings.default`

TODO:
- [x] Add tests

srid reviewed (commented):
Nice. Quick review.

- [x] A separate `./test/*/flake.nix` would be good for testing defaults.
- [x] Rebase new `master` (#276), and add documentation in `docs/defaults.md`
  - Let's elaborate the examples in the Markdown docs, and have the option `description` link to them.
- [ ] Not sure about the `defaults.settings.default-foo` naming; let me think about it.

srid commented:
cf. https://github.com/srid/haskell-flake/discussions/301

shivaraj-bh commented:
added tests: https://github.com/srid/haskell-flake/pull/275/commits/18581246feea93c7d37c02787ad879d1c110b1db

srid reviewed (commented):
My CI is down, but running `nixci build` locally fails:

```
$ nixci build https://github.com/srid/haskell-flake/pull/275
[..]
$ nix log /nix/store/n1qm2m8qmg3d7q25swl3h8cqxhsng9na-haskell-flake-test-0.1.0.0.drv | tail -6
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/n1qm2m8qmg3d7q25swl3h8cqxhsng9na-haskell-flake-test-0.1.0.0.drv^*'
Configuring haskell-flake-test-0.1.0.0...
CallStack (from HasCallStack):
  withMetadata, called at libraries/Cabal/Cabal/src/Distribution/Simple/Utils.hs:368:14 in Cabal-3.10.1.0:Distribution.Simple.Utils
Error: Setup: Encountered missing or private dependencies:
hello
```

shivaraj-bh commented:
> My CI is down, but running nixci build locally fails:

I fixed and tested on Linux.

`hello` wasn’t a great example to use anyway, it isn’t a library package. I switched over to `random` package.

srid commented on nix/modules/project/settings/default.nix:
```diff
@@ -88,6 +96,6 @@ in
           );
       in
-      lib.mapAttrs applySettingsFor project.config.settings;
+      lib.mapAttrs applySettingsFor (lib.recursiveUpdate emptyPackageSettings project.config.settings);
```
Why do we need `lib.recursiveUpdate emptyPackageSetting`?

shivaraj-bh replied:
It considers the packages that are mentioned in `haskellProjects.<prj-name>.packages` but not in `haskellProjects.<prj-name>.settings`. i should include a test that would cover this case. Also, is the name misleading? Anything better you suggest?

srid replied:
The `settings` option already does this:

https://github.com/srid/haskell-flake/blob/847292fc793a5c15c873e52e7751ee4267ef32a0/nix/modules/project/settings/default.nix#L21-L27

So, `project.config.settings` is already polyfilled with packages from `project.config.packages`.

srid reviewed (commented):
LGTM, just not sure what `emptyPackageSettings` is for exactly. When I replace it with `{}`, the test still passes.
